Compilation issue with MySQL-5.7 sources

[ 10%] Building CXX object storage/q4m/CMakeFiles/queue.dir/src/ha_queue.cc.o
/usr/src/mysql/mysql-5.7/storage/q4m/src/ha_queue.cc:43:22: fatal error: sql_priv.h: No such file or directory
compilation terminated.
storage/q4m/CMakeFiles/queue.dir/build.make:86: recipe for target 'storage/q4m/CMakeFiles/queue.dir/src/ha_queue.cc.o' failed
make[2]: *** [storage/q4m/CMakeFiles/queue.dir/src/ha_queue.cc.o] Error 1
CMakeFiles/Makefile2:748: recipe for target 'storage/q4m/CMakeFiles/queue.dir/all' failed
make[1]: *** [storage/q4m/CMakeFiles/queue.dir/all] Error 2
Makefile:160: recipe for target 'all' failed
make: *** [all] Error 2
root@cloud:/usr/src/mysql/mysql-5.7#

--- a/README
+++ b/README
@@ -3,7 +3,7 @@ Q4M - a Message Queue for MySQL
 http://q4m.github.io/
 
 Q4M is a message queue that works as a pluggable storage engine of MySQL
-5.1 / 5.5 / 5.6.
+5.1 / 5.5 / 5.6 / 5.7.
 
 * MySQL 5.6 Compatibility
 

--- a/src/ha_queue.cc
+++ b/src/ha_queue.cc
@@ -88,6 +88,7 @@ extern uint build_table_filename(char *buff, size_t bufflen, const char *db,
 #endif
 
 #include "ha_queue.h"
+#include "queue_mysql_compat.h"
 #include "adler32.c"
 
 BOOST_STATIC_ASSERT(sizeof(queue_file_header_t)
@@ -109,6 +110,10 @@ using namespace std;
   ((all) >= COMPACT_THRESHOLD && (free) * 4 >= (all) * 3)
 #define Q4M ".Q4M"
 #define Q4T ".Q4T"
+
+#ifdef Q4M_HAVE_PSI_MEMORY_KEY
+static PSI_memory_key queue_key_memory_queue_share;
+#endif
 
 static ulonglong mmap_max;
 static MYSQL_SYSVAR_ULONGLONG(mmap_max,
@@ -368,7 +373,7 @@ queue_row_t *queue_row_t::create_checksum(const iovec* iov, int iovcnt)
   }
   
   queue_row_t *row =
-    static_cast<queue_row_t*>(my_malloc(checksum_size(), MYF(0)));
+    static_cast<queue_row_t*>(q4m_my_malloc(checksum_size(), MYF(0)));
   create_checksum(row, sz, adler);
   
   return row;
@@ -680,7 +685,7 @@ queue_share_t *queue_share_t::get_share(const char *table_name, bool if_is_open)
   }
   
   /* alloc */
-  if (my_multi_malloc(MYF(MY_WME | MY_ZEROFILL), &share, sizeof(queue_share_t),
+  if (q4m_my_multi_malloc(MYF(MY_WME | MY_ZEROFILL), &share, sizeof(queue_share_t),
 		      &tmp_name, table_name_length + 1, NullS)
       == NULL) {
     goto ERR_RETURN;
@@ -2655,7 +2660,7 @@ int ha_queue::prepare_rows_buffer(size_t sz)
     while (rows_reserved < sz) {
       rows_reserved *= 2;
     }
-    if ((rows = static_cast<uchar*>(my_malloc(rows_reserved, MYF(0))))
+    if ((rows = static_cast<uchar*>(q4m_my_malloc(rows_reserved, MYF(0))))
 	== NULL) {
       return -1;
     }
@@ -2665,7 +2670,7 @@ int ha_queue::prepare_rows_buffer(size_t sz)
       new_reserve *= 2;
     } while (new_reserve < rows_size + sz);
     void *pt;
-    if ((pt = my_realloc(rows, new_reserve, MYF(0))) == NULL) {
+    if ((pt = q4m_my_realloc(rows, new_reserve, MYF(0))) == NULL) {
       return -1;
     }
     rows = static_cast<uchar*>(pt);
@@ -2859,17 +2864,42 @@ static bool show_status(handlerton *hton, THD *thd, stat_print_fn *print,
   }
 }
 
+#ifdef Q4M_HAVE_PSI_MEMORY_KEY
+static PSI_memory_info all_queue_memory[]=
+{
+  { &queue_key_memory_queue_share, "queue_share", 0}
+};
+
+void init_queue_psi_keys()
+{
+  const char* category= "queue";
+  int count;
+
+  count= array_elements(all_queue_memory);
+  mysql_memory_register(category, all_queue_memory, count);
+}
+#endif /* Q4M_HAVE_PSI_MEMORY_KEY */
+
 static int init_plugin(void *p)
 {
   queue_hton = (handlerton *)p;
-  
+
+#ifdef Q4M_HAVE_PSI_MEMORY_KEY
+  init_queue_psi_keys();
+#endif
+
   pthread_mutex_init(&open_mutex, MY_MUTEX_INIT_FAST);
   pthread_mutex_init(&listener_mutex, MY_MUTEX_INIT_FAST);
   pthread_mutex_init(&tbl_stat_mutex, MY_MUTEX_INIT_FAST);
   pthread_mutex_init(&compile_expr_mutex, MY_MUTEX_INIT_FAST);
   pthread_mutex_init(&stat_mutex, MY_MUTEX_INIT_FAST);
   my_hash_init(&queue_open_tables, system_charset_info, 32, 0, 0,
-	    reinterpret_cast<my_hash_get_key>(queue_share_t::get_share_key), 0, 0);
+	    reinterpret_cast<my_hash_get_key>(queue_share_t::get_share_key),
+#ifdef Q4M_HAVE_PSI_MEMORY_KEY
+        0, 0, queue_key_memory_queue_share);
+#else
+        0, 0);
+#endif
   queue_hton->state = SHOW_OPTION_YES;
   queue_hton->close_connection = queue_connection_t::close;
   queue_hton->create = create_handler;

--- a/src/ha_queue.cc
+++ b/src/ha_queue.cc
@@ -58,10 +58,6 @@ extern uint build_table_filename(char *buff, size_t bufflen, const char *db,
 #undef PACKAGE_VERSION
 #include <mysql/plugin.h>
 
-#if MYSQL_VERSION_ID >= 50500
-#define my_free(PTR, FLAG) my_free(PTR)
-#endif
-
 #if MYSQL_VERSION_ID < 50500
 #define mysql_mutex_lock(mutex) pthread_mutex_lock(mutex)
 #define mysql_mutex_unlock(mutex) pthread_mutex_unlock(mutex)
@@ -818,7 +814,7 @@ queue_share_t *queue_share_t::get_share(const char *table_name, bool if_is_open)
   thr_lock_delete(&share->store_lock);
   pthread_rwlock_destroy(&share->rwlock);
   pthread_mutex_destroy(&share->compact_mutex);
-  my_free(reinterpret_cast<uchar*>(share), MYF(0));
+  q4m_my_free(reinterpret_cast<uchar*>(share), MYF(0));
  ERR_RETURN:
   pthread_mutex_unlock(&open_mutex);
   return NULL;
@@ -873,7 +869,7 @@ void queue_share_t::release()
     thr_lock_delete(&store_lock);
     pthread_rwlock_destroy(&rwlock);
     pthread_mutex_destroy(&compact_mutex);
-    my_free(reinterpret_cast<uchar*>(this), MYF(0));
+    q4m_my_free(reinterpret_cast<uchar*>(this), MYF(0));
   }
   
   pthread_mutex_unlock(&open_mutex);
@@ -1620,7 +1616,7 @@ int queue_share_t::writer_do_append(append_list_t *l)
       if (i - writev_from >= IOV_MAX
 	  || writev_len + i->iov_len > SSIZE_MAX / 2) {
 	if (sys_writev(fd, &*writev_from, i - writev_from) != writev_len) {
-	  my_free(iov[0].iov_base, MYF(0));
+	  q4m_my_free(iov[0].iov_base, MYF(0));
 	  return HA_ERR_CRASHED_ON_USAGE;
 	}
 	writev_from = i;
@@ -1629,7 +1625,7 @@ int queue_share_t::writer_do_append(append_list_t *l)
       writev_len += i->iov_len;
     }
     if (sys_writev(fd, &*writev_from, iov.end() - writev_from) != writev_len) {
-      my_free(iov[0].iov_base, MYF(0));
+      q4m_my_free(iov[0].iov_base, MYF(0));
       return HA_ERR_CRASHED_ON_USAGE;
     }
     sync_file(fd);
@@ -1656,7 +1652,7 @@ int queue_share_t::writer_do_append(append_list_t *l)
                                   + bytes_total_add);
   }
   
-  my_free(iov[0].iov_base, MYF(0));
+  q4m_my_free(iov[0].iov_base, MYF(0));
   return 0;
 }
 
@@ -2685,7 +2681,7 @@ void ha_queue::free_rows_buffer(bool force)
     return;
   }
   if (rows != NULL) {
-    my_free(rows, MYF(0));
+    q4m_my_free(rows, MYF(0));
     rows = NULL;
     rows_size = 0;
   }

--- a/src/ha_queue.cc
+++ b/src/ha_queue.cc
@@ -2780,7 +2780,11 @@ static queue_share_t* get_share_check(const char* db_table_name)
     db = buf;
     tbl = tbl + 1;
   } else {
+#if MYSQL_VERSION_ID >= 50705
+    db = current_thd->db().str;
+#else
     db = current_thd->db;
+#endif
     tbl = db_table_name;
   }
   if (db == NULL) {

--- a/src/ha_queue.cc
+++ b/src/ha_queue.cc
@@ -690,7 +690,11 @@ queue_share_t *queue_share_t::get_share(const char *table_name, bool if_is_open)
   /* init members that would always succeed in doing so */
   share->ref_cnt = 1;
   share->table_name = tmp_name;
+#if MYSQL_VERSION_ID >= 50703
+  my_stpcpy(share->table_name, table_name);
+#else
   strmov(share->table_name, table_name);
+#endif
   share->table_name_length = table_name_length;
   pthread_mutex_init(&share->compact_mutex, MY_MUTEX_INIT_FAST);
   {

--- a/src/ha_queue.cc
+++ b/src/ha_queue.cc
@@ -3146,7 +3146,11 @@ my_bool queue_wait_init(UDF_INIT *initid, UDF_ARGS *args, char *message)
   {
     THD* thd = current_thd;
     if (thd->lex != NULL
+#if MYSQL_VERSION_ID >= 50705
+        && thd->lex->select_lex->table_list.elements != 0) {
+#else
         && thd->lex->select_lex.table_list.elements != 0) {
+#endif
       strcpy(message, "as of MySQL 5.6, the function cannot be used within a WHERE clause");
       return 1;
     }

--- a/src/ha_queue.cc
+++ b/src/ha_queue.cc
@@ -40,7 +40,9 @@ extern uint build_table_filename(char *buff, size_t bufflen, const char *db,
 				 const char *table, const char *ext,
 				 uint flags);
 #else
+#if MYSQL_VERSION_ID < 50700
 #include "sql_priv.h"
+#endif
 #include "sql_base.h"
 #include "sql_table.h"
 #include "probes_mysql.h"

--- a/src/ha_queue.cc
+++ b/src/ha_queue.cc
@@ -21,6 +21,7 @@ extern "C" {
 #include <sys/types.h>
 #include <sys/uio.h>
 #include <unistd.h>
+#include <sys/time.h>
 }
 #include <algorithm>
 #include <functional>

--- a/src/queue_mysql_compat.h
+++ b/src/queue_mysql_compat.h
@@ -1,0 +1,20 @@
+#pragma once
+
+#if defined(HAVE_PSI_INTERFACE) && MYSQL_VERSION_ID >= 50706
+#  define Q4M_HAVE_PSI_MEMORY_KEY
+#endif
+
+#ifdef Q4M_HAVE_PSI_MEMORY_KEY
+#  define q4m_my_malloc(size, flags) \
+  my_malloc(queue_key_memory_queue_share, size, flags)
+#  define q4m_my_realloc(ptr, size, flags) \
+  my_realloc(queue_key_memory_queue_share, ptr, size, flags)
+#  define q4m_my_multi_malloc(flags, ...) \
+  my_multi_malloc(queue_key_memory_queue_share, flags, __VA_ARGS__)
+#else
+#  define q4m_my_malloc(size, flags) my_malloc(size, flags)
+#  define q4m_my_realloc(ptr, size, flags) \
+  my_realloc(ptr, size, flags)
+#  define q4m_my_multi_malloc(flags, ...) \
+  my_multi_malloc(flags, __VA_ARGS__)
+#endif

--- a/src/queue_mysql_compat.h
+++ b/src/queue_mysql_compat.h
@@ -18,3 +18,7 @@
 #  define q4m_my_multi_malloc(flags, ...) \
   my_multi_malloc(flags, __VA_ARGS__)
 #endif
+
+#if MYSQL_VERSION_ID >= 50500
+#define q4m_my_free(PTR, FLAG) my_free(PTR)
+#endif

--- a/t/05-multireader-read.c
+++ b/t/05-multireader-read.c
@@ -67,7 +67,8 @@ int main(int argc, char **argv)
   }
   for (i = 0; i < loop;) {
     /* read data and print */
-    res = do_select("SELECT * FROM q4m_t WHERE queue_wait('q4m_t')");
+    res = do_select("SELECT queue_wait('q4m_t')");
+    res = do_select("SELECT * FROM q4m_t");
     if (mysql_num_rows(res) != 0) {
       printf("%s\n", mysql_fetch_row(res)[0]);
       i++;

--- a/t/05-multireader.t
+++ b/t/05-multireader.t
@@ -78,7 +78,10 @@ for (my $i = 0; $i < $NUM_CHILDREN; $i++) {
             my $rows;
             do {
                 $rows = $dbh->selectall_arrayref(
-                    "select * from q4m_t where queue_wait('q4m_t')",
+                    "select queue_wait('q4m_t')",
+                ) or die $dbh->errstr;
+                $rows = $dbh->selectall_arrayref(
+                    "select * from q4m_t",
                 ) or die $dbh->errstr;
             } while (! @$rows);
             print "$rows->[0]->[0]\n";

--- a/t/05-multirw-core.c
+++ b/t/05-multirw-core.c
@@ -89,7 +89,8 @@ int main(int argc, char **argv)
     insert_row(i + start_value, var_length);
     /* queue_wait */
     while (1) {
-      res = do_select("SELECT * FROM q4m_t WHERE queue_wait('q4m_t')");
+      res = do_select("SELECT queue_wait('q4m_t')");
+      res = do_select("SELECT * FROM q4m_t");
       if (mysql_num_rows(res) != 0) {
 	break;
       }

--- a/t/05-multirw.t
+++ b/t/05-multirw.t
@@ -70,7 +70,10 @@ for (my $i = 0; $i < $NUM_CHILDREN; $i++) {
             my $rows;
             do {
                 $rows = $dbh->selectall_arrayref(
-                    "select * from q4m_t where queue_wait('q4m_t')",
+                    "select queue_wait('q4m_t')",
+                ) or die $dbh->errstr;
+                $rows = $dbh->selectall_arrayref(
+                    "select * from q4m_t",
                 ) or die $dbh->errstr;
             } while (@$rows == 0);
             print "$rows->[0]->[0]\n";

--- a/t/05-multiwait-core.c
+++ b/t/05-multiwait-core.c
@@ -87,7 +87,8 @@ int main(int argc, char **argv)
   for (i = 0; i < loop; i++) {
     /* read data and print */
     while (1) {
-      res = do_select("SELECT * FROM q4m_t WHERE queue_wait('q4m_t')");
+      res = do_select("SELECT queue_wait('q4m_t')");
+      res = do_select("SELECT * FROM q4m_t");
       if (mysql_num_rows(res) != 0) {
 	break;
       }

--- a/t/05-multiwait.t
+++ b/t/05-multiwait.t
@@ -69,7 +69,10 @@ for (my $i = 0; $i < $NUM_CHILDREN; $i++) {
             my $rows;
             do {
                 $rows = $dbh->selectall_arrayref(
-                    "select * from q4m_t where queue_wait('q4m_t')",
+                    "select queue_wait('q4m_t')",
+                ) or die $dbh->errstr;
+                $rows = $dbh->selectall_arrayref(
+                    "select * from q4m_t",
                 ) or die $dbh->errstr;
             } while (@$rows == 0);
             print "$rows->[0]->[0]\n";

